### PR TITLE
Uniform image distribution

### PR DESCRIPTION
**Pull request by yudistrange**
- add `count` column in the images table
- modify the `images/random` api to send images with least `count`
- update the `count` whenever a pair of image is voted on
   - Note: The skipped pair of images is not counted.

### CODE REVIEW

**Comment by loleg** [comment]
Does that query in the API really work? I mean, you're picking a random item from an ordered list, it's still random... Please explain :-)

Also, if the images are no longer truly random, please rename (or add separately) the API call as "get_next_pair".

One more thought: if the sort order was randomly shuffled, then sequentially retrieved (POPped), we could do away with a count field. I'm not comfortable with DB row updates adding additional contention and risking performance issues in the app.

**Comment by loleg** [comment]
If skipped images are not counted, they will appear more often, which is not want we want. Please count them as well.

**Comment by loleg** [comment]
To my last point: let's store the counts in memory, using a slim lookup table. We don't care about this data, not really (we have analytics for that), and that would assuage my performance concerns.

**Comment by yudistrange** [comment]
If we store the count data in memory, it would be transient and get reset after every deploy. I suppose we are fine with that, right?

**Comment by loleg** [comment]
While in our testing the code is stable, and I'm happy with the minimally-invasive option taken now, we still have the issue that without having an actual images-shown count, we have no idea what we are measuring. Every time we restart the server, the in-memory variable gets reset. Therefore, before this actually has benefit we need to add a procedure which initializes the image map from the votes database.

I am also slightly concerned by potential impact of our code change to the production deployment - memory limitations, specifically. Even just storing the ID and count, the new code is a potential scalability issue once you have millions of image records. If you agree, we should add this as a note to the production section of the README.

**Comment by loleg** [comment]
I've also rolled a dependency refresh into this, so we can bring the environments up to date.

**Comment by loleg** [comment]
- It made no sense to initialize the counter in `manage.py`, since that happens outside of the application context.
- I've refactored the code of the now-called `image_tracker` significantly for clarity, and because it wasn't giving consistent results.
- Test suite and local testing are passing.

**Comment by loleg** [comment]
In the last set of improvements, I've cached the data in the application context. I don't think this is a significant improvement over the global variable. And does contribute to our memory usage. But it is a more Pythonic (or Flask-onic) way of doing things. See https://flask.palletsprojects.com/en/1.1.x/appcontext/

File: streetwise/api/images.py [hunk 1]
```diff
@@ -46,12 +56,16 @@ class ImageRandom(Resource):
     @ns.doc('random_images')
     @ns.marshal_list_with(ImageModel)
     def get(self, campaign_id):
-        if campaign_id is None or campaign_id == 'null':
-            print("Null campaign requested: selecting default (1)")
-            campagn_id = 1
+        # Prime the image counter if needed
+        if not 'image_counter' in g:
+            g.image_counter = init_image_counter()
```
Comment (loleg) [inline streetwise/api/images.py:61]: This is rather "buried", and I would have preferred to have it happen in the init of the app, but I couldn't get it working in `__init__.py` - so here it is. Some unlucky user may get a timeout if they open the un-primed app and the init takes a long time.

File: streetwise/api/images.py [hunk 2]
```diff
@@ -3,17 +3,27 @@
...
+from .helper.image_tracker import (
+    init_image_counter, least_displayed_images
+)
+
 ns = api_rest.namespace('image',
     description = 'Image operations'
 )
 
+# How many images to randomly select before counting
+IMAGES_RANDOM_WALK = 50
```
Comment (loleg) [inline streetwise/api/images.py:22]: Why 10? Why not 50 or even 100? A higher number leads to a better distribution, but slower performance.

File: tests/test_images.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        db.session.commit()
+
+        image_tracker.init_image_counter()
+        # print(image_tracker.IMAGE_COUNTER_DICT)
+
+        resp = client.get('/api/image/random/%d' % campaign1.id)
+        assert resp.status_code == 201
+
+        result = json.loads(resp.data)
+        assert images[0].id in (result[0]['id'], result[1]['id'])
```
Comment (loleg) [inline tests/test_images.py:43]: This is a very minimal test. A more sophisticated one would statistically test the distribution of the image counter after a bunch of random hits.

**Comment by loleg** [comment]
I'm getting this error:
```
  File "/home/oleg/Clients/Streetwise/app/streetwise/api/helper/image_tracker.py", line 59, in sort_images_by_display_count
    if IMAGE_COUNTER_DICT[None].get(image.id) is None:
KeyError: None
```

**Comment by yudistrange** [comment]
I have reverted the change to handle empty campaign. The migration to fix our db should take care of that. 

I think you are getting the key error because in your local db there is no vote without a campaign id

**Pull request merged**